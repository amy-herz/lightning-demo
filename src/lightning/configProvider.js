import { getDefaultConfig } from './defaultConfig';

// Closure to hold the APIs if and when available
let PROVIDED_IMPL = getDefaultConfig();
let FORM_FACTOR;

let configProvided = false;

export default function configProviderService(serviceAPI) {
  if (!configProvided) {
    PROVIDED_IMPL = {
      getCoreInfo:
        serviceAPI.getInitializer &&
        serviceAPI.getInitializer('coreInfoConfig'),
      getPathPrefix: serviceAPI.getPathPrefix,
      getFormFactor: serviceAPI.getFormFactor,
      getToken: serviceAPI.getToken,
      getLocale: serviceAPI.getLocale,
      getLocalizationService: serviceAPI.getLocalizationService,
      sanitizeDOM: serviceAPI.sanitizeDOM
    };
    configProvided = true;
    FORM_FACTOR = undefined;
  } else {
    throw new Error(
      'ConfigProvider can only be set once at initilization time'
    );
  }

  return { name: 'lightning-config-provider' };
}

export function getPathPrefix() {
  return (
    (PROVIDED_IMPL &&
      PROVIDED_IMPL.getPathPrefix &&
      PROVIDED_IMPL.getPathPrefix()) ||
    ''
  );
}

export function getToken(name) {
  return (
    PROVIDED_IMPL && PROVIDED_IMPL.getToken && PROVIDED_IMPL.getToken(name)
  );
}

export function getLocale() {
  return PROVIDED_IMPL && PROVIDED_IMPL.getLocale && PROVIDED_IMPL.getLocale();
}

export function getFormFactor() {
  if (!FORM_FACTOR) {
    if (PROVIDED_IMPL && PROVIDED_IMPL.getFormFactor) {
      FORM_FACTOR = PROVIDED_IMPL.getFormFactor();
    } else {
      FORM_FACTOR = 'DESKTOP';
    }
  }

  return FORM_FACTOR;
}

export function getLocalizationService() {
  return (
    PROVIDED_IMPL &&
    PROVIDED_IMPL.getLocalizationService &&
    PROVIDED_IMPL.getLocalizationService()
  );
}

export function sanitizeDOM(dirty, config) {
  if (PROVIDED_IMPL && PROVIDED_IMPL.sanitizeDOM) {
    return PROVIDED_IMPL.sanitizeDOM(dirty, config);
  }
  return '';
}

export function getCoreInfo() {
  return (
    (PROVIDED_IMPL && PROVIDED_IMPL.getCoreInfo) || {
      untrustedContentDomain: '.a.forceusercontent.com',
      localhostPort: '',
      securePort: ''
    }
  );
}
