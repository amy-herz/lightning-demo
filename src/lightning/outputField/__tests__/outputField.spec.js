import { createElement } from 'lwc';
import Element from 'lightning/outputField';
import store from './mockData.json';
// eslint-disable-next-line lwc/no-aura-libs
import { utils as localeUtils } from 'lightning:IntlLibrary';

const createOutputField = fieldName => {
    const element = createElement('lightning-output-field', { is: Element });
    element.fieldName = fieldName;
    document.body.appendChild(element);
    element.wireRecordUi(store);
    return element;
};

function setLocaleMock(locale) {
    localeUtils.getLocaleTag = () => locale;
}

function waitForReady(element) {
    let pollCount = 0;
    function pollClassList(resolve, reject) {
        if (pollCount > 10) {
            // poll ten times on the microtask queue
            reject('timed out waiting for ready');
        } else if (element.classList.contains('slds-hide')) {
            pollCount++;
            // eslint-disable-next-line lwc/no-set-timeout
            setTimeout(pollClassList(resolve, reject), 0);
        } else {
            resolve();
        }
    }

    return new Promise((resolve, reject) => {
        pollClassList(resolve, reject);
    });
}

const fields = {
    NameSimple: {
        label: 'Simple Name',
        value: 'Simple Name',
    },
};

/**
 * Help Text generates a new, different ID, each time, so any jest testing
 * would need to be able to ignore different help text IDs
 */
describe('lightning-output-field', () => {
    it('text', () => {
        const element = createOutputField('Text');
        return waitForReady(element).then(() => {
            return expect(element).toMatchSnapshot();
        });
    });

    it('text with field reference', () => {
        const element = createOutputField({
            fieldApiName: 'Text',
            objectApiName: 'entity',
        });
        return waitForReady(element).then(() => {
            return expect(element).toMatchSnapshot();
        });
    });

    it('throws an error on objectApiName mismatch', () => {
        const element = createElement('lightning-output-field', {
            is: Element,
        });
        const data = JSON.parse(JSON.stringify(store));
        element.fieldName = {
            fieldApiName: 'Text',
            objectApiName: 'notEntity',
        };
        document.body.appendChild(element);
        expect(() => {
            element.wireRecordUi(data);
        }).toThrowErrorMatchingSnapshot();
    });

    it('noLabel', () => {
        const element = createOutputField('Text');
        element.variant = 'label-hidden';
        return waitForReady(element).then(() => {
            return expect(element).toMatchSnapshot();
        });
    });

    it('nonexistent', async () => {
        const element = createOutputField('NonField');
        return Promise.resolve().then(() => {
            return expect(element).toMatchSnapshot();
        });
    });

    it('boolean', () => {
        const element = createOutputField('Boolean');
        return waitForReady(element).then(() => {
            return expect(element).toMatchSnapshot();
        });
    });

    it('email', () => {
        const element = createOutputField('Email');
        return waitForReady(element).then(() => {
            return expect(element).toMatchSnapshot();
        });
    });

    it('phone', () => {
        const element = createOutputField('Phone');
        return waitForReady(element).then(() => {
            return expect(element).toMatchSnapshot();
        });
    });

    it('picklist', () => {
        const element = createOutputField('Picklist');
        return waitForReady(element).then(() => {
            return expect(element).toMatchSnapshot();
        });
    });

    it('richtext fallback', () => {
        const element = createOutputField('RichText');
        return waitForReady(element).then(() => {
            return expect(element).toMatchSnapshot();
        });
    });

    it('url', () => {
        const element = createOutputField('Url');
        return waitForReady(element).then(() => {
            return expect(element).toMatchSnapshot();
        });
    });

    it('adds slds-hide to invalid or hidden field', () => {
        const element = createOutputField('InvalidField');
        return expect(waitForReady(element)).rejects.toBeDefined();
    });

    describe('formatted-name', () => {
        it('Default Name', () => {
            const element = createOutputField('Name');
            return waitForReady(element).then(() => {
                return expect(element).toMatchSnapshot();
            });
        });
        it('Name based on locale', () => {
            setLocaleMock('ja_JP');
            const element = createOutputField('Name');
            return waitForReady(element).then(() => {
                return expect(element).toMatchSnapshot();
            });
        });
        it('renders simple names as text field', () => {
            const element = createOutputField('NameSimple');
            return waitForReady(element).then(() => {
                const textElement = element.shadowRoot.querySelector(
                    'lightning-formatted-text'
                );
                expect(textElement).toBeDefined();
                expect(textElement.shadowRoot.textContent).toEqual(
                    fields.NameSimple.value
                );
            });
        });
    });

    describe('formatted-address', () => {
        it('Default Address', () => {
            const element = createOutputField('BillingAddress');
            return waitForReady(element).then(() => {
                return expect(element).toMatchSnapshot();
            });
        });

        it('format address based on locale', () => {
            setLocaleMock('cn-CN');
            const element = createOutputField('BillingAddress');
            return waitForReady(element).then(() => {
                return expect(element).toMatchSnapshot();
            });
        });
    });
});
