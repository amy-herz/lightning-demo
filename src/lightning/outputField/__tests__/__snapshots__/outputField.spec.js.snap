// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lightning-output-field boolean 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Checkbox label
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-input
        class="slds-form-element__static slds-form-element"
      >
        #shadow-root(open)
          <span
            class="slds-assistive-text"
            data-aria="true"
          />
          <div
            class="slds-form-element__control"
          >
            <span
              class="slds-checkbox"
            >
              <input
                disabled=""
                id="checkbox-11"
                name="Boolean"
                type="checkbox"
                value=""
              />
              <label
                class="slds-checkbox__label"
                for="checkbox-11"
              >
                <span
                  class="slds-checkbox_faux"
                />
                <span
                  class="slds-form-element__label slds-assistive-text"
                >
                  Checkbox label
                </span>
              </label>
            </span>
          </div>
      </lightning-input>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field email 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Email
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-email
        class="slds-form-element__static"
      >
        #shadow-root(open)
          <a
            href="mailto:the@email.something"
          >
            <lightning-icon
              class="slds-icon-utility-email slds-icon_container"
            >
              #shadow-root(open)
                <lightning-primitive-icon>
                  #shadow-root(open)
                    <svg
                      aria-hidden="true"
                      class="slds-icon slds-icon-text-default slds-icon_x-small"
                      data-key="email"
                      focusable="false"
                    >
                      <use
                        xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#email"
                      />
                    </svg>
                </lightning-primitive-icon>
                <span
                  class="slds-assistive-text"
                >
                  Email
                </span>
            </lightning-icon>
             
            the@email.something
          </a>
      </lightning-formatted-email>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field formatted-address Default Address 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Address
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-address
        class="slds-form-element__static"
      >
        #shadow-root(open)
          <a
            href="https://www.google.com/maps/?q=11%20Farm%20Avenue%0AHartford,%20AK%0AUS,%206156"
            rel="noopener"
            target="_blank"
            title="11 Farm Avenue
Hartford, AK
US, 6156"
          >
            <div
              class="slds-truncate"
            >
              11 Farm Avenue
            </div>
            <div
              class="slds-truncate"
            >
              Hartford, AK
            </div>
            <div
              class="slds-truncate"
            >
              US, 6156
            </div>
          </a>
      </lightning-formatted-address>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field formatted-address format address based on locale 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Address
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-address
        class="slds-form-element__static"
      >
        #shadow-root(open)
          <a
            href="https://www.google.com/maps/?q=US%0AAK,%20Hartford%0A11%20Farm%20Avenue,%206156"
            rel="noopener"
            target="_blank"
            title="US
AK, Hartford
11 Farm Avenue, 6156"
          >
            <div
              class="slds-truncate"
            >
              US
            </div>
            <div
              class="slds-truncate"
            >
              AK, Hartford
            </div>
            <div
              class="slds-truncate"
            >
              11 Farm Avenue, 6156
            </div>
          </a>
      </lightning-formatted-address>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field formatted-name Default Name 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Full Name
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-name
        class="slds-form-element__static"
      >
        #shadow-root(open)
          Mr. Jim  Steele  
      </lightning-formatted-name>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field formatted-name Name based on locale 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Full Name
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-name
        class="slds-form-element__static"
      >
        #shadow-root(open)
          Steele  Jim  
      </lightning-formatted-name>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field noLabel 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-text
        class="slds-form-element__static"
      >
        #shadow-root(open)
          Some Plain Text
      </lightning-formatted-text>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field nonexistent 1`] = `
<lightning-output-field
  class="slds-form-element slds-hide"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    />
    <div
      class="slds-form-element__control"
    >
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field phone 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Phone Number
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-phone
        class="slds-form-element__static"
      >
        #shadow-root(open)
          <a
            href="tel:403-203-1048"
          >
            403-203-1048
          </a>
      </lightning-formatted-phone>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field picklist 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Picklist
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-text
        class="slds-form-element__static"
      >
        #shadow-root(open)
          PicklistDisplayValue
      </lightning-formatted-text>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field richtext fallback 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Rich Text
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-rich-text
        class="slds-form-element__static slds-rich-text-editor__output"
      >
        #shadow-root(open)
          <span>
            &lt;img src="/servlet/servlet.FileDownload?file=01570000000Q6Ep" alt="Red" border="0"/&gt;
          </span>
      </lightning-formatted-rich-text>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field text 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Plain Text
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-text
        class="slds-form-element__static"
      >
        #shadow-root(open)
          Some Plain Text
      </lightning-formatted-text>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field text with field reference 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Plain Text
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-text
        class="slds-form-element__static"
      >
        #shadow-root(open)
          Some Plain Text
      </lightning-formatted-text>
      <slot />
    </div>
</lightning-output-field>
`;

exports[`lightning-output-field throws an error on objectApiName mismatch 1`] = `"objectApiName (notEntity) for field Text does not match the objectApiName provided for the form (entity)."`;

exports[`lightning-output-field url 1`] = `
<lightning-output-field
  class="slds-form-element"
>
  #shadow-root(open)
    <span
      class="slds-form-element__label"
    >
      Url
    </span>
    <div
      class="slds-form-element__control"
    >
      <lightning-formatted-url
        class="slds-form-element__static"
      >
        #shadow-root(open)
          <a
            href="http://www.google.com"
          >
            http://www.google.com
          </a>
      </lightning-formatted-url>
      <slot />
    </div>
</lightning-output-field>
`;
